server/src/main/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverService.java

```diff
@@ -281,6 +290,63 @@ private RolloverResult rolloverDataStream(
             return new RolloverResult(newWriteIndexName, originalWriteIndex.getName(), currentState);
         }
 
+        AtomicReference<DataStreamAutoShardingEvent> newAutoShardingEvent = new AtomicReference<>(dataStream.getAutoShardingEvent());
+        if (autoShardingResult != null) {
+            switch (autoShardingResult.type()) {
+                case NO_CHANGE_REQUIRED -> {
```

Comment:
I think we can avoid the `AtomicReference` here by having the switch expression yield the event. For example, something like (pseudocode):

```java
DataStreamAutoShardingEvent newAutoShardingEvent;
if (autoShardingResult != null) {
    newAutoShardingEvent = switch (autoShardingResult.type()) {
        case NO_CHANGE_REQUIRED -> {
            …
            yield dataStream.getAutoShardingEvent();
        }
        case INCREASE_SHARDS, DECREASE_SHARDS -> {
            …
            yield new DataStreamAutoShardingEvent(…);
        }
        …
    }
}
```

Reply:
++